File: src/main/java/com/hazza/LeetCodeByJava/SegmentTwo/No123BestTimeToBuyAndSellStockIII.java
package com.hazza.LeetCodeByJava.SegmentTwo;

import org.omg.CORBA.INTERNAL;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/description/
 * User: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 18-2-5
 * Time: 3:43 PM
 */
public class No123BestTimeToBuyAndSellStockIII {

    // solution 1
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length < 2) return 0;
        int len = prices.length;
        int[] dp = new int[len];

        int k = 2; // the number of transactions at most
        for (int i = 0; i < k; i++) {
            int tmpMax = -prices[0];
            for (int j = 1; j < len; j++) {
                int prev = dp[j];
                dp[j] = Math.max(dp[j - 1], prices[j] + tmpMax);
                tmpMax = Math.max(tmpMax, prev - prices[j]);
            }
        }

        return dp[len - 1];
    }

    // solution 2
    public int maxProfit1(int[] prices) {
        if (prices == null || prices.length < 2) return 0;
        int buy1 = Integer.MIN_VALUE, buy2 = Integer.MIN_VALUE, sell1 = 0, sell2 = 0;

        for (int price: prices) {
            sell2 = Math.max(sell2, buy2 + price);
            buy2 = Math.max(buy2, sell1 - price);
            sell1 = Math.max(sell1, buy1 + price);
            buy1 = Math.max(buy1, -price);
        }

        return sell2;
    }
    
}

File: src/main/python/leetcode_by_python/dp/No123_Best_Time_to_Buy_and_Sell_Stock_III.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-11-18
import sys
from typing import List


class Solution:
    def maxProfit1(self, prices: List[int]) -> int:
        if not prices:
            return 0

        prices_len = len(prices)
        dp = [0] * prices_len

        for _ in range(2):
            tmp_max = -prices[0]
            for i in range(1, prices_len):
                prev = dp[i]
                dp[i] = max(dp[i - 1], prices[i] + tmp_max)
                tmp_max = max(tmp_max, prev - prices[i])

        return dp[prices_len - 1]

    def maxProfit2(self, prices: List[int]) -> int:
        if not prices:
            return 0

        buy1, buy2 = -sys.maxsize, -sys.maxsize
        sell1, sell2 = 0, 0

        for p in prices:
            sell2 = max(sell2, buy2 + p)
            buy2 = max(buy2, sell1 - p)
            sell1 = max(sell1, buy1 + p)
            buy1 = max(buy1, -p)

        return sell2

